File: api-schema/hub/references.ts
import type { Handle } from "./hub-users.js";

export type ReferenceNominationState =
	| "nominated"
	| "accepted"
	| "declined"
	| "submitted"
	| "expired";

export type ReferenceInboxRequestKind = "to_nominate" | "to_respond";

export interface ReferenceQuestion {
	question_id: string;
	text: string;
	min_chars: number;
	max_chars: number;
	required: boolean;
}

export interface HubReferenceRequestSummary {
	kind: ReferenceInboxRequestKind;
	request_id: string;
	nomination_id?: string;
	org_domain: string;
	org_name: string;
	opening_title: string;
	candidate_handle?: Handle;
	max_references?: number;
	questions: ReferenceQuestion[];
	response_deadline: string;
	state?: ReferenceNominationState;
	created_at: string;
}

export interface ListReferenceRequestsIncomingRequest {
	filter_kind?: ReferenceInboxRequestKind[];
	filter_state?: ReferenceNominationState[];
	pagination_key?: string;
	limit?: number;
}

export interface ListReferenceRequestsIncomingResponse {
	requests: HubReferenceRequestSummary[];
	next_pagination_key?: string;
}

export interface NominateReferencesRequest {
	request_id: string;
	nominee_handles: Handle[];
}

export interface AcceptReferenceNominationRequest {
	nomination_id: string;
}

export interface DeclineReferenceNominationRequest {
	nomination_id: string;
}

export interface SubmitReferenceResponseRequest {
	nomination_id: string;
	answers: Array<{
		question_id: string;
		response_text: string;
	}>;
}

export interface ValidationError {
	field: string;
	message: string;
}

export function validateListReferenceRequestsIncomingRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (r.limit !== undefined) {
		if (typeof r.limit !== "number" || r.limit < 1 || r.limit > 100) {
			errors.push({ field: "limit", message: "Must be between 1 and 100" });
		}
	}

	if (r.pagination_key !== undefined && typeof r.pagination_key !== "string") {
		errors.push({ field: "pagination_key", message: "Must be a string" });
	}

	return errors;
}

export function validateNominateReferencesRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.request_id !== "string" || r.request_id.trim() === "") {
		errors.push({ field: "request_id", message: "Must be a non-empty string" });
	}

	if (!Array.isArray(r.nominee_handles)) {
		errors.push({ field: "nominee_handles", message: "Must be an array" });
	} else {
		if (r.nominee_handles.length < 1 || r.nominee_handles.length > 5) {
			errors.push({
				field: "nominee_handles",
				message: "Must have 1-5 items",
			});
		}
		for (const h of r.nominee_handles) {
			if (typeof h !== "string" || h.trim() === "") {
				errors.push({
					field: "nominee_handles",
					message: "All handles must be non-empty strings",
				});
				break;
			}
		}
	}

	return errors;
}

export function validateAcceptReferenceNominationRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.nomination_id !== "string" || r.nomination_id.trim() === "") {
		errors.push({
			field: "nomination_id",
			message: "Must be a non-empty string",
		});
	}

	return errors;
}

export function validateDeclineReferenceNominationRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.nomination_id !== "string" || r.nomination_id.trim() === "") {
		errors.push({
			field: "nomination_id",
			message: "Must be a non-empty string",
		});
	}

	return errors;
}

export function validateSubmitReferenceResponseRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.nomination_id !== "string" || r.nomination_id.trim() === "") {
		errors.push({
			field: "nomination_id",
			message: "Must be a non-empty string",
		});
	}

	if (!Array.isArray(r.answers)) {
		errors.push({ field: "answers", message: "Must be an array" });
	} else {
		for (const answer of r.answers) {
			if (!answer || typeof answer !== "object") {
				errors.push({
					field: "answers",
					message: "All answers must be objects",
				});
				break;
			}
			const a = answer as Record<string, unknown>;
			if (typeof a.question_id !== "string" || a.question_id.trim() === "") {
				errors.push({
					field: "answers.question_id",
					message: "Must be a non-empty string",
				});
				break;
			}
			if (typeof a.response_text !== "string") {
				errors.push({
					field: "answers.response_text",
					message: "Must be a string",
				});
				break;
			}
		}
	}

	return errors;
}

File: api-schema/hub/references.go
package hub

import (
	"vetchium-api-server.typespec/common"
)

type ReferenceNominationState string

const (
	ReferenceNominationStateNominated ReferenceNominationState = "nominated"
	ReferenceNominationStateAccepted  ReferenceNominationState = "accepted"
	ReferenceNominationStateDeclined  ReferenceNominationState = "declined"
	ReferenceNominationStateSubmitted ReferenceNominationState = "submitted"
	ReferenceNominationStateExpired   ReferenceNominationState = "expired"
)

type ReferenceInboxRequestKind string

const (
	ReferenceInboxRequestKindToNominate ReferenceInboxRequestKind = "to_nominate"
	ReferenceInboxRequestKindToRespond  ReferenceInboxRequestKind = "to_respond"
)

type ReferenceQuestion struct {
	QuestionID string `json:"question_id"`
	Text       string `json:"text"`
	MinChars   int32  `json:"min_chars"`
	MaxChars   int32  `json:"max_chars"`
	Required   bool   `json:"required"`
}

type HubReferenceRequestSummary struct {
	Kind             ReferenceInboxRequestKind `json:"kind"`
	RequestID        string                    `json:"request_id"`
	NominationID     *string                   `json:"nomination_id,omitempty"`
	OrgDomain        string                    `json:"org_domain"`
	OrgName          string                    `json:"org_name"`
	OpeningTitle     string                    `json:"opening_title"`
	CandidateHandle  *string                   `json:"candidate_handle,omitempty"`
	MaxReferences    *int32                    `json:"max_references,omitempty"`
	Questions        []ReferenceQuestion       `json:"questions"`
	ResponseDeadline string                    `json:"response_deadline"`
	State            *ReferenceNominationState `json:"state,omitempty"`
	CreatedAt        string                    `json:"created_at"`
}

type ListReferenceRequestsIncomingRequest struct {
	FilterKind    []ReferenceInboxRequestKind `json:"filter_kind,omitempty"`
	FilterState   []ReferenceNominationState  `json:"filter_state,omitempty"`
	PaginationKey *string                     `json:"pagination_key,omitempty"`
	Limit         *int32                      `json:"limit,omitempty"`
}

type ListReferenceRequestsIncomingResponse struct {
	Requests          []HubReferenceRequestSummary `json:"requests"`
	NextPaginationKey *string                      `json:"next_pagination_key,omitempty"`
}

type NominateReferencesRequest struct {
	RequestID      string   `json:"request_id"`
	NomineeHandles []string `json:"nominee_handles"`
}

type AcceptReferenceNominationRequest struct {
	NominationID string `json:"nomination_id"`
}

type DeclineReferenceNominationRequest struct {
	NominationID string `json:"nomination_id"`
}

type ReferenceAnswer struct {
	QuestionID   string `json:"question_id"`
	ResponseText string `json:"response_text"`
}

type SubmitReferenceResponseRequest struct {
	NominationID string            `json:"nomination_id"`
	Answers      []ReferenceAnswer `json:"answers"`
}

func (r ListReferenceRequestsIncomingRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.Limit != nil && (*r.Limit < 1 || *r.Limit > 100) {
		errs = append(errs, common.ValidationError{Field: "limit", Message: "Must be between 1 and 100"})
	}
	return errs
}

func (r NominateReferencesRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.RequestID == "" {
		errs = append(errs, common.ValidationError{Field: "request_id", Message: "Must be a non-empty string"})
	}
	if len(r.NomineeHandles) < 1 || len(r.NomineeHandles) > 5 {
		errs = append(errs, common.ValidationError{Field: "nominee_handles", Message: "Must have 1-5 items"})
	}
	for _, h := range r.NomineeHandles {
		if h == "" {
			errs = append(errs, common.ValidationError{Field: "nominee_handles", Message: "All handles must be non-empty strings"})
			break
		}
	}
	return errs
}

func (r AcceptReferenceNominationRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.NominationID == "" {
		errs = append(errs, common.ValidationError{Field: "nomination_id", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r DeclineReferenceNominationRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.NominationID == "" {
		errs = append(errs, common.ValidationError{Field: "nomination_id", Message: "Must be a non-empty string"})
	}
	return errs
}

func (r SubmitReferenceResponseRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.NominationID == "" {
		errs = append(errs, common.ValidationError{Field: "nomination_id", Message: "Must be a non-empty string"})
	}
	if len(r.Answers) == 0 {
		errs = append(errs, common.ValidationError{Field: "answers", Message: "Must have at least one answer"})
	}
	for i, ans := range r.Answers {
		if ans.QuestionID == "" {
			errs = append(errs, common.ValidationError{Field: "answers[" + string(rune(i)) + "].question_id", Message: "Must be a non-empty string"})
		}
		if ans.ResponseText == "" {
			errs = append(errs, common.ValidationError{Field: "answers[" + string(rune(i)) + "].response_text", Message: "Must be a non-empty string"})
		}
	}
	return errs
}
